Sources/Sentry/SentryHttpInterceptor.m

```diff
@@ -0,0 +1,112 @@
+#import "SentryHttpInterceptor+Private.h"
+#import "SentryHub+Private.h"
+#import "SentrySDK+Private.h"
+#import "SentryScope+Private.h"
+#import "SentryTraceHeader.h"
+
+@interface
+SentryHttpInterceptor () <NSURLSessionDelegate>
+
+@property (nonatomic, strong) NSURLSession *session;
+
++ (void)configureSessionConfiguration:(NSURLSessionConfiguration *)configuration;
+
+@end
+
+@implementation SentryHttpInterceptor
+
++ (void)configureSessionConfiguration:(NSURLSessionConfiguration *)configuration
+{
+    if (configuration == nil)
+        return;
+
+    NSMutableArray *protocolClasses = configuration.protocolClasses != nil
+        ? [NSMutableArray arrayWithArray:[configuration protocolClasses]]
+        : [[NSMutableArray alloc] init];
+
+    if (![protocolClasses containsObject:[self class]]) {
+        [protocolClasses insertObject:[self class] atIndex:0];
+    }
+
+    configuration.protocolClasses = protocolClasses;
+}
+
++ (BOOL)canInitWithRequest:(NSURLRequest *)request
+{
+    // Intercept the request if it is a http/https request
+    // not targeting Sentry and there is transaction in the scope.
+    NSNumber *intercepted = [NSURLProtocol propertyForKey:SENTRY_INTERCEPTED_REQUEST
+                                                inRequest:request];
+    if (intercepted != nil && [intercepted boolValue])
+        return false;
+
+    NSURL *apiUrl = [NSURL URLWithString:SentrySDK.options.dsn];
+    if ([request.URL.host isEqualToString:apiUrl.host] &&
+        [request.URL.path containsString:apiUrl.path])
+        return false;
+
+    if (SentrySDK.currentHub.scope.span == nil)
+        return false;
+
+    return ([request.URL.scheme isEqualToString:@"http"] ||
+        [request.URL.scheme isEqualToString:@"https"]);
```

Comment:
Where is `canInitWithRequest` used?

If the decision to do something relies on a `span` being on the scope we have to remember that after this line even before returning, this might not be true anymore. Meaning we might require synchronizing things if it requires atomicity. Assuming it doesn't then the caller just needs to make sure `span` is there before using it and not relying on this check alone.

Leaving the null check last and defaulting to `true` to be clear we're 'disallow-listing' here:

```suggestion
    if (![request.URL.scheme isEqualToString:@"http"] &&
        ![request.URL.scheme isEqualToString:@"https"]))
        return false;
        
    if (SentrySDK.currentHub.scope.span == nil)
        return false;

    return true;
```

Reply:
canInitRequest is used by the System to determine whether or protocol can handle the request, we have no control where this is called.

Reply:
We perform a second check in the moment we need to add the header in the request.